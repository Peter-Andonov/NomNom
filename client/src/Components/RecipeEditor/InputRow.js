import React from 'react';
import styled from 'styled-components';


const StyledInput = styled.input`
    font-size: 1.2rem;
    width: 5rem;
    padding: 0.5rem;
    border-radius: 5px;
    border: 2px solid black;
`;

const StyledSelect = styled.select`
    font-size: 1.2rem;
    width: 12rem;
    padding: 0.5rem;
    border-radius: 5px;
    border: 2px solid black;
`;

const InputRow = (props) => {

    const setQuantityInput = (e) => {
        props.setQuantityInputValues(e.target.value, props.idx);
    }

    const setUnitInput = (e) => {
        props.setUnitInputValues(e.target.value, props.idx);
    }

    const setIngredientInput = (e) => {
        props.setIngredientInputValues(e.target.value, props.idx);
    }

    const unitOptions = props.units.map((unit) => <option key={unit._id} value={unit._id} >{unit.name}</option>)

    const ingredientOptions = props.ingredients.map((ingredient) => <option key={ingredient._id} value={ingredient._id} >{ingredient.name}</option>)

    return (
        
        <tr>
            <td>
                <StyledInput value={props.quantityValue} onChange={setQuantityInput} type='number' />
            </td>
            <td>
                <StyledSelect value={props.unitValue} onChange={setUnitInput}>
                    <option value="" hidden >Select Unit</option>
                    {unitOptions}
                </StyledSelect>
            </td>
            <td>
                <StyledSelect value={props.ingredientValue} onChange={setIngredientInput}>
                    <option value="" hidden >Select Ingredient</option>
                    {ingredientOptions}
                </StyledSelect>
            </td>
        </tr>
    );
};

export default InputRow;