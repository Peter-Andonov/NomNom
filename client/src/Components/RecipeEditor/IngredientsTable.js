import React, { useState } from 'react';
import InputRow from './InputRow';
import styled from 'styled-components';
import addIcon from '../../Images/Icons/add-24px.svg';
import removeIcon from '../../Images/Icons/remove-24px.svg';


const Wrapper = styled.div`
    background-color: rgb(205, 255, 204);
    padding: 1.5rem;
    margin: 1.5rem;
    border-radius: 20px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
`;

const Container = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
`;

const StyledInput = styled.input`
    font-size: 1.2rem;
    width: 50%;
    padding: 0.5rem;
    border-radius: 5px;
    border: 2px solid black;
`;

const SectionIcon = styled.img`
    margin-left: 0.5rem;
    &:hover{
        cursor: pointer;
    }
`;

const ColHeading = styled.td`
    text-align: center;
`;

const IngredientsTable = (props) => {

    const [inputFields, setInputFields] = useState(['input-0']);

    const addInput = () => {
        const newInput = `input-${inputFields.length}`;
        setInputFields([...inputFields, newInput]);
    };

    const removeInput = () => {

        if (inputFields.length === 1) {
            return;
        }
        //remove the last input field from the state
        const newInputFields = [...inputFields];
        newInputFields.splice((newInputFields.length - 1), 1);
        setInputFields(newInputFields);

        //remove the last quantity input value from the state
        const newQuantityInputValues = [...props.sectionState.quantities];
        newQuantityInputValues.splice((newQuantityInputValues.length - 1), 1);
        props.sectionState.quantities = newQuantityInputValues;

        //remove the last unit input value from the state
        const newUnitInputValues = [...props.sectionState.units];
        newUnitInputValues.splice((newUnitInputValues.length - 1), 1);
        props.sectionState.units = newUnitInputValues;

        //remove the last ingredient input value from the state
        const newInngredientInputValues = [...props.sectionState.ingredients];
        newInngredientInputValues.splice((newInngredientInputValues.length - 1), 1);
        props.sectionState.ingredients = newInngredientInputValues;
    };

    const handleNameInput = (value) => {
        const { id, quantities, units, ingredients } = { ...props.sectionState };
        const newName = value;
        props.handleInput({
            id,
            name: newName,
            quantities,
            units,
            ingredients
        });
    }

    const handleQuantityInputValues = (value, idx) => {
        const { id, name, quantities, units, ingredients } = { ...props.sectionState };
        const newQuantities = [...quantities];
        newQuantities[idx] = value;
        props.handleInput({
            id,
            name,
            quantities: newQuantities,
            units,
            ingredients
        });
    };

    const handleUnitInputValues = (value, idx) => {
        const { id, name, quantities, units, ingredients } = { ...props.sectionState };
        const newUnits = [...units];
        newUnits[idx] = value;
        props.handleInput({
            id,
            name,
            quantities,
            units: newUnits,
            ingredients
        });
    };

    const handleIngredientInputValues = (value, idx) => {
        const { id, name, quantities, units, ingredients } = { ...props.sectionState };
        const newIngredients = [...ingredients];
        newIngredients[idx] = value;
        props.handleInput({
            id,
            name,
            quantities,
            units,
            ingredients: newIngredients
        });
    };


    return (
        <Wrapper>
            <StyledInput
                placeholder='Section name'
                type='text'
                id={props.sectionState.id}
                value={props.sectionState.name}
                onChange={(e) => handleNameInput(e.target.value)}
            />
            <table>
                <thead>
                    <tr>
                        <ColHeading>Quantity</ColHeading>
                        <ColHeading>Unit</ColHeading>
                        <ColHeading>Ingredient</ColHeading>
                    </tr>
                </thead>
                <tbody>
                    {inputFields.map((input, idx) =>
                        <InputRow
                            key={input}
                            idx={idx}
                            units={props.units}
                            ingredients={props.ingredients}
                            setQuantityInputValues={handleQuantityInputValues}
                            setUnitInputValues={handleUnitInputValues}
                            setIngredientInputValues={handleIngredientInputValues}
                        />)}
                </tbody>
            </table>
            <Container>
                <SectionIcon src={addIcon} onClick={addInput} />
                <SectionIcon src={removeIcon} onClick={removeInput} />
            </Container>
        </Wrapper>
    );
};


export default IngredientsTable;