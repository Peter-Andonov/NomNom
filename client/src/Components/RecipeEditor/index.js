import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import Axios from 'axios';
import TextEditor from '../TextEditor';
import ImageSelector from '../ImageSelector';
import Input from '../RegisterForm/Input';
import IngredientsTable from './IngredientsTable';
import addIcon from '../../Images/Icons/add-24px.svg';
import removeIcon from '../../Images/Icons/remove-24px.svg';


const Wrapper = styled.div`
    background-color: white;
    height: auto;
    width: 80%;
    display: flex;
    flex-direction: column;
    align-items: center;
`;

const Container = styled.div`
    margin-top: 0.5rem;
    width: 100%;
    height: auto;
    display: flex;
    flex-wrap: wrap;
    align-items: left;
    justify-content: center;
`;

const Column = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
`;

const ErrorMessage = styled.div`
    font-size: 1rem;
    color: red;
`;

const SectionIcon = styled.img`
    margin-left: 0.5rem;
    &:hover{
        cursor: pointer;
    }
`;

const StyledInput = styled.input`
    font-size: 1.2rem;
    width: 6.5rem;
    padding: 0.5rem;
    border-radius: 5px;
    border: 2px solid black;
`;

const StyledSelect = styled.select`
    font-size: 1.2rem;
    width: 7.8rem;
    padding: 0.5rem;
    border-radius: 5px;
    border: 2px solid black;
`;

const InfoLabel = styled.div`
    padding-left: 0.5rem;
    width: 12rem;
`;

const Button = styled.button`
    margin-top: 1.5rem;
    margin-bottom: 1.5rem;
    padding-left: 1rem;
    padding-top: 0.5rem;
    padding-right: 1rem;
    padding-bottom: 0.5rem;
    background: rgb(255, 151, 23);
    border: none;
    border-radius: 20px;
    color: white;
    font-family: 'Sriracha', cursive;
    font-size: 1.5rem;
    &:focus {
        outline: none;
    };
    &:hover {
        cursor: pointer;
    };
`;


const RecipeEditor = (props) => {

    const [allUnits, setAllUnits] = useState([]);
    const [allIngredients , setAllIngredients] = useState([]);


    useEffect(() => {
        Axios(`http://localhost:5000/api/ingredient/all`, {
            method: "GET",
            params: {
                sortCrit: 'name',
                sortOrd: 'asc'
            }
        }).then((res) => {
            setAllIngredients(res.data.ingredients);
        }).catch((err) => {
            console.log(err);
        });
    }, []);

    useEffect(() => {
        Axios(`http://localhost:5000/api/unit/all`, {
            method: "GET",
            params: {
                sortCrit: 'name',
                sortOrd: 'asc'
            }
        }).then((res) => {
            setAllUnits(res.data.units);
        }).catch((err) => {
            console.log(err);
        });
    }, []);

    const addIngredientSection = () => {
        const newIngredientSection = `section-${props.ingredientSections.length}`;
        props.setIngredientSections([...props.ingredientSections, {
            _id: newIngredientSection,
            name: '',
            quantities: [''],
            units: [''],
            ingredients: ['']
        }]);
    };

    const removeIngredientSection = () => {

        if (props.ingredientSections.length === 1) {
            return;
        };

        //remove the last ingredient section from the state
        const newIngredientSections = [...props.ingredientSections];
        newIngredientSections.splice((newIngredientSections.length - 1), 1);
        props.setIngredientSections(newIngredientSections);
    };

    const addInputRow = (idx) => {
        const newState = [...props.ingredientSections];

        newState[idx].quantities.push('')
        newState[idx].units.push('')
        newState[idx].ingredients.push('')
        props.setIngredientSections(newState);
    };

    const removeInputRow = (idx) => {

        const newState = [...props.ingredientSections];

        if (newState[idx].ingredients.length <= 1) {
            return;
        };

        newState[idx].quantities.splice((newState[idx].quantities.length - 1), 1);
        newState[idx].units.splice((newState[idx].units.length - 1), 1);
        newState[idx].ingredients.splice((newState[idx].ingredients.length - 1), 1);

        props.setIngredientSections(newState);
    };

    const handleInput = (newSectionState) => {
        const newIngredientSections = props.ingredientSections.map((section) => {
            if (section._id === newSectionState._id) {
                return newSectionState
            }
            return section;
        });
        props.setIngredientSections(newIngredientSections);
    };

    return (
        <Wrapper>
            <h1>Create Recipe</h1>
            <h3>Recipe Title</h3>
            <Input
                value={props.title}
                onChange={props.setTitle} />
            <h3>Cover Image</h3>
            <ImageSelector
                imageUrl={props.coverImageUrl}
                setImageUrl={props.setCoverImageUrl}
            />
            <h3>Short Description</h3>
            <TextEditor
                editorState={props.shortDescriptionState}
                setEditorState={props.setShortDescriptionState}
            />
            <h3>Steps to create</h3>
            <TextEditor
                editorState={props.stepsState}
                setEditorState={props.setStepsState}
            />
            <Container>
                <h3>Ingredient Sections</h3>
                <SectionIcon src={addIcon} onClick={addIngredientSection} />
                <SectionIcon src={removeIcon} onClick={removeIngredientSection} />
            </Container>
            <Container>
                {props.ingredientSections && props.ingredientSections.map((section, idx) =>
                    <IngredientsTable
                        key={section._id}
                        idx={idx}
                        units={allUnits}
                        ingredients={allIngredients}
                        sectionState={props.ingredientSections[idx]}
                        handleInput={handleInput}
                        addInputRow={addInputRow}
                        removeInputRow={removeInputRow}
                    />)}
            </Container>
            <h3>Additional recipe information</h3>
            <Column>
                <Container>
                    <InfoLabel>Time to prepare</InfoLabel>
                    <StyledInput
                        type='number'
                        value={props.prepTime}
                        onChange={(e) => props.setPrepTime(e.target.value)}
                    />
                    <InfoLabel>min</InfoLabel>
                </Container>
                <Container>
                    <InfoLabel>Time to cook</InfoLabel>
                    <StyledInput
                        type='number'
                        value={props.cookTime}
                        onChange={(e) => props.setCookTime(e.target.value)}
                    />
                    <InfoLabel>min</InfoLabel>
                </Container>
                <Container>
                    <InfoLabel>Serves</InfoLabel>
                    <StyledInput
                        type='number'
                        value={props.serves}
                        onChange={(e) => props.setServes(e.target.value)}
                    />
                    <InfoLabel>people</InfoLabel>
                </Container>
                <Container>
                    <InfoLabel>Difficulty</InfoLabel>
                    <StyledSelect
                    value={props.difficulty}
                    onChange={(e) => props.setDifficulty(e.target.value)}>
                        <option value="" hidden >Difficulty</option>
                        <option value="easy" >Easy</option>
                        <option value="medium" >Medium</option>
                        <option value="hard" >Hard</option>
                    </StyledSelect>
                    <InfoLabel></InfoLabel>
                </Container>
            </Column>
            {props.error && <ErrorMessage>{props.errorMessage}</ErrorMessage>}
            <Button onClick={props.action} >Save recipe</Button>
        </Wrapper>
    );
};


export default RecipeEditor;