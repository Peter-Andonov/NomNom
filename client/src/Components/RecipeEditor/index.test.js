import React from 'react'
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TestSetup from '../../Utils/TestUtils/test-utils';


import RecipeEditor from './index';

jest.mock('./IngredientsTable.js', () => () => <div />);
jest.mock('../TextEditor/index.js', () => () => <div />);

describe('RecipeEditor component', () => {
    it("renders correctly", () => {
        const { asFragment } = render(
            <TestSetup value={{
                loggedIn: true,
                user: {
                    _id: '1234',
                    email: "john.doe@gmail.com",
                    role: 'admin',
                    firstName: undefined,
                    lastName: undefined,
                    profilePicUrl: '',
                    favouriteRecipes: []
                }
            }}>
                <RecipeEditor />
            </TestSetup>);
        expect(asFragment()).toMatchSnapshot();
    });
});