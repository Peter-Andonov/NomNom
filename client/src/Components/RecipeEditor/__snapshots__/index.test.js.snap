// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders guest navigation 1`] = `
<DocumentFragment>
  <div
    class="sc-fzozJi hhEBat"
  >
    <h1>
      Create Recipe
    </h1>
    <h3>
      Recipe Title
    </h3>
    <div
      class="sc-AxgMl hmIBnY"
    >
      <label
        class="sc-AxheI uxhxw"
      />
      <input
        class="sc-Axmtr ityRYL"
        value=""
      />
    </div>
    <h3>
      Cover Image
    </h3>
    <div
      class="sc-AxhUy gPrrHp"
    >
      <div>
        <input
          accept="image/*"
          class="sc-AxjAm htLRco"
          id="file"
          name="File"
          placeholder="Upload image"
          type="file"
        />
        <label
          class="sc-AxirZ dymrJU"
          for="file"
        >
          + Select Image
        </label>
      </div>
    </div>
    <h3>
      Short Description
    </h3>
    <div />
    <h3>
      Steps to create
    </h3>
    <div />
    <div
      class="sc-fzoLsD hIIKPn"
    >
      <h3>
        Ingredient Sections
      </h3>
      <img
        class="sc-fznyAO IgzWd"
        src="add-24px.svg"
      />
      <img
        class="sc-fznyAO IgzWd"
        src="remove-24px.svg"
      />
    </div>
    <div
      class="sc-fzoLsD hIIKPn"
    />
    <h3>
      Additional recipe information
    </h3>
    <div
      class="sc-fzpans fzlHkI"
    >
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Time to prepare
        </div>
        <input
          class="sc-fznKkj enPeTG"
          type="number"
          value=""
        />
        <div
          class="sc-fzokOt gDEqsa"
        >
          min
        </div>
      </div>
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Time to cook
        </div>
        <input
          class="sc-fznKkj enPeTG"
          type="number"
          value=""
        />
        <div
          class="sc-fzokOt gDEqsa"
        >
          min
        </div>
      </div>
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Serves
        </div>
        <input
          class="sc-fznKkj enPeTG"
          type="number"
          value=""
        />
        <div
          class="sc-fzokOt gDEqsa"
        >
          people
        </div>
      </div>
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Difficulty
        </div>
        <select
          class="sc-fznZeY hNvtzG"
        >
          <option
            hidden=""
            value=""
          >
            Difficulty
          </option>
          <option
            value="easy"
          >
            Easy
          </option>
          <option
            value="medium"
          >
            Medium
          </option>
          <option
            value="hard"
          >
            Hard
          </option>
        </select>
        <div
          class="sc-fzokOt gDEqsa"
        />
      </div>
    </div>
    <button
      class="sc-fzqBZW fajpxm"
    >
      Save recipe
    </button>
  </div>
</DocumentFragment>
`;

exports[`RecipeEditor component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-fzozJi hhEBat"
  >
    <h1>
      Create Recipe
    </h1>
    <h3>
      Recipe Title
    </h3>
    <div
      class="sc-AxgMl hmIBnY"
    >
      <label
        class="sc-AxheI uxhxw"
      />
      <input
        class="sc-Axmtr ityRYL"
        value=""
      />
    </div>
    <h3>
      Cover Image
    </h3>
    <div
      class="sc-AxhUy gPrrHp"
    >
      <div>
        <input
          accept="image/*"
          class="sc-AxjAm htLRco"
          id="file"
          name="File"
          placeholder="Upload image"
          type="file"
        />
        <label
          class="sc-AxirZ dymrJU"
          for="file"
        >
          + Select Image
        </label>
      </div>
    </div>
    <h3>
      Short Description
    </h3>
    <div />
    <h3>
      Steps to create
    </h3>
    <div />
    <div
      class="sc-fzoLsD hIIKPn"
    >
      <h3>
        Ingredient Sections
      </h3>
      <img
        class="sc-fznyAO IgzWd"
        src="add-24px.svg"
      />
      <img
        class="sc-fznyAO IgzWd"
        src="remove-24px.svg"
      />
    </div>
    <div
      class="sc-fzoLsD hIIKPn"
    />
    <h3>
      Additional recipe information
    </h3>
    <div
      class="sc-fzpans fzlHkI"
    >
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Time to prepare
        </div>
        <input
          class="sc-fznKkj enPeTG"
          type="number"
          value=""
        />
        <div
          class="sc-fzokOt gDEqsa"
        >
          min
        </div>
      </div>
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Time to cook
        </div>
        <input
          class="sc-fznKkj enPeTG"
          type="number"
          value=""
        />
        <div
          class="sc-fzokOt gDEqsa"
        >
          min
        </div>
      </div>
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Serves
        </div>
        <input
          class="sc-fznKkj enPeTG"
          type="number"
          value=""
        />
        <div
          class="sc-fzokOt gDEqsa"
        >
          people
        </div>
      </div>
      <div
        class="sc-fzoLsD hIIKPn"
      >
        <div
          class="sc-fzokOt gDEqsa"
        >
          Difficulty
        </div>
        <select
          class="sc-fznZeY hNvtzG"
        >
          <option
            hidden=""
            value=""
          >
            Difficulty
          </option>
          <option
            value="easy"
          >
            Easy
          </option>
          <option
            value="medium"
          >
            Medium
          </option>
          <option
            value="hard"
          >
            Hard
          </option>
        </select>
        <div
          class="sc-fzokOt gDEqsa"
        />
      </div>
    </div>
    <button
      class="sc-fzqBZW fajpxm"
    >
      Save recipe
    </button>
  </div>
</DocumentFragment>
`;
